### PR TITLE
fix tests and other tweaks related to node 20

### PR DESCRIPTION
**Pull request by lperson**
This PR fixes a couple of failing tests.

Additional context in a [slack thread](https://larryperson.slack.com/archives/C06ET56J36U/p1711142002791949?thread_ts=1710944581.531959&cid=C06ET56J36U).

# Description

## ERR_REQUIRE_ESM errors

Pinned `strip-ansi` to 6.0.1.

## Failing actions

* Updated `actions/checkout@v3` to `actions/checkout@v4` so it doesn't complain about node 20.
* Updated `actions/cache@v3` to `actions/cache@v4` so it doesn't complain about node 20.

## Failure in `__test__/backend.test.js`

Specify the campaign due date as milliseconds since the epoch when testing with SQLite.

I did a quick smoke test of copying a campaign in the UI when using SQLite and no additional special handling was necessary.

## Failure in `__test__/extensions/action-handlers/ngpvan-action.test.js`

The failing test was testing a failure caused by trying to parse an object as if it were a JSON string.

It’s failing because the exception thrown by a JSON parse error changed between Node versions.

### Node 16.18.0

```
==> node
Welcome to Node.js v16.18.0.
Type ".help" for more information.
> JSON.parse("{")
Uncaught SyntaxError: Unexpected end of JSON input
> JSON.parse({})
Uncaught SyntaxError: Unexpected token o in JSON at position 1
```

### Node 20.11.1

```
==> nvm use 20.11.1
Now using node v20.11.1 (npm v10.2.4)
==> node
Welcome to Node.js v20.11.1.
Type ".help" for more information.
> JSON.parse({})
Uncaught SyntaxError: "[object Object]" is not valid JSON

```

# Checklist:

- [ ] I have manually tested my changes on desktop and mobile
- [ ] The test suite passes locally with my changes
- [ ] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [ ] [My change is 300 lines of code or less](https://github.com/StateVoicesNational/Spoke/blob/main/CONTRIBUTING.md#submitting-your-pull-request), or has a documented reason in the description why it’s longer
- [ ] I have made any necessary changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] My PR is labeled [WIP] if it is in progress

### CODE REVIEW

File: __test__/backend.test.js [hunk 1]
```diff
@@ -632,30 +636,19 @@ describe("graphql test suite", () => {
           typeof copiedCampaign.due_by === "number" ||
           typeof copiedCampaign.due_by === "string"
         ) {
-          let parsedDate = new Date(copiedCampaign.due_by);
+          const parsedDate = new Date(copiedCampaign.due_by);
```
Comment (lperson) [inline __test__/backend.test.js:639]: To silence the linter.

File: __test__/backend.test.js [hunk 2]
```diff
@@ -632,30 +636,19 @@ describe("graphql test suite", () => {
...
-          }
+          expect(copiedCampaign.due_by).toEqual(campaign.due_by);
         }
+
         if (
           typeof copiedCampaign.features === "object" &&
           copiedCampaign.features
         ) {
-          let jsonString = JSON.stringify(copiedCampaign.features);
+          const jsonString = JSON.stringify(copiedCampaign.features);
```
Comment (lperson) [inline __test__/backend.test.js:651]: To silence the linter.

**Comment by lperson** [comment]
@mau11  I think it's worth reviewing this and merging it if you approve. It gets the tests a lot further than before.

I'll continue to work on the redis tests failing and make another PR.

**Review by mau11 — APPROVED** [review]
thanks Larry 🎉 great fixes!

**Comment by engelhartrueben** [comment]
This looks great!

**Pull request merged**